### PR TITLE
Uuidd and lastlog2 tmpfiles fixes

### PR DESCRIPTION
**Pull request by eworm-de**

### CODE REVIEW

**Comment by karelzak** [comment]
The directory /var/lib/libuuid has to be persistent. It's not a temporary file.

The temporary file is @runstatedir@/uuidd/request (see misc-utils/uuidd.socket.in); on Fedora, we have this directory in tmpfiles.d, maybe we can do the same for upstream.

**Comment by eworm-de** [comment]
Well, yes... All directories in `/var/lib` should be persistent,  anyway `tmpfiles` can be *misused* to (re-)create them.

The issue I am trying to address is this one: [uuidd doesn't start, /var/lib/libuuid is missing](https://gitlab.archlinux.org/archlinux/packaging/packages/util-linux/-/issues/5)

The directory is not created when building and installing with `meson`. (No idea about `autotools`...)

**Comment by t-8ch** [comment]
IMO creating files in `/var` via `tmpfiles` is completely fine.
It allows deleting `/var` for factory resets/immutable systems as per https://0pointer.net/blog/projects/stateless.html

**Comment by karelzak** [comment]
The directory is used to avoid duplicate UUIDs, you really don't want any reset there ;-) Is it possible with tmpfiles? (just create and don't touch anymore).

* uuidd.service.in executes uuidd with uudd user and group, the same should be used for tmpfiles

* Fedora uses tmpfile for /run/uuidd
```
cat uuidd-tmpfiles.conf 
d /run/uuidd 2775 uuidd uuidd
```
I guess we can add this also to upstream.

* we need solution for non-systemd distros, there is misc-utils/uuidd.rc.in, mkdir and chown would be probably enough

* we can /var/lib/libuuid to the tmpfile (with proper user and group) if it's possible to set it persistent

**Comment by eworm-de** [comment]
The directory is not touched if it exists (with the correct permissions if given). So this is created once and stays persistent.

**Comment by eworm-de** [comment]
Hmm, `misc-utils/uuidd.rc.in` handles `@runstatedir@/uuidd` already. Is that `/run/uuidd` or `/var/lib/libuuid`?

**Comment by eworm-de** [comment]
Pushed another change... I guess using `@localstatedir@/lib/libuuid` is the best bet.

File: misc-utils/uuidd.conf.in [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# This file is part of uuidd.
+#
+# See tmpfiles.d(5) for details
+#
+d /run/uuidd 2775 uuidd uuidd -
+d /var/lib/libuuid 0755 uuidd uuidd -
```
Comment (karelzak) [inline misc-utils/uuidd.conf.in:6]: You can also use @runstatedir@ and @localstatedir@ in this template.

Comment (eworm-de) [reply misc-utils/uuidd.conf.in:6]: Ah, good point... Updating in a moment.

**Comment by karelzak** [comment]
We also need to install misc-utils/uuidd.conf.in by autotools, which means adding the file to misc-utils/Makemodule.am as tmpfiles_DATA += misc-utils/uuidd.conf.

I'd also suggest renaming the file from uuidd.conf to uuidd-tmpfiles.conf

File: misc-utils/uuidd.conf.in (hunk 1 above)
Comment (eworm-de) [reply misc-utils/uuidd.conf.in]: Well, that fails. We do not have `@localstatedir@` with `meson`, no? The result of `@runstatedir@` is as expected, though.

**Comment by eworm-de** [comment]
> I'd also suggest renaming the file from uuidd.conf to uuidd-tmpfiles.conf

Just the template or the final file?

**Comment by karelzak** [comment]
> > I'd also suggest renaming the file from uuidd.conf to uuidd-tmpfiles.conf
> 
> Just the template or the final file?

Both? It's confusing, the daemon does not have any config file :-)

**Comment by eworm-de** [comment]
> Both? It's confusing, the daemon does not have any config file :-)

Should we do the same for `lastlog2.conf` then?

**Comment by karelzak** [comment]
> > Both? It's confusing, the daemon does not have any config file :-)
> 
> Should we do the same for `lastlog2.conf` then?

Good point, I think yes.

File: misc-utils/uuidd-tmpfiles.conf.in [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# This file is part of uuidd.
+#
+# See tmpfiles.d(5) for details
+#
+d @runstatedir@/uuidd 2775 uuidd uuidd -
+d /var/lib/libuuid 0755 uuidd uuidd -
```
Comment (karelzak) [inline misc-utils/uuidd-tmpfiles.conf.in:6]: @localstatedir@ rather than /var/lib/libuuid ?

**Comment by karelzak** [comment]
I guess we can merge this also to stable/v2.40 to have it in the next v2.40. I'd like to have -rc2 tomorrow or on Friday.

File: misc-utils/uuidd.conf.in (hunk 1 above)
Comment (karelzak) [reply misc-utils/uuidd.conf.in]: Ah, now I see your comment about @localstatedir@. Is there anything else used in meson?

Comment (karelzak) [reply misc-utils/uuidd.conf.in]: I'll try to fix it.

**Comment by karelzak** [comment]
@eworm-de, there is a patch to introduce localstatedir in autotools and meson:
https://github.com/eworm-de/util-linux/pull/1/commits/2e04d96f43801de064cdbcb41508a56544dd5ad5

**Comment by eworm-de** [comment]
> I guess we can merge this also to stable/v2.40 to have it in the next v2.40.

Yes, please. 👍
Parts of this would even make sense in *stable/V2.39*...

> I'd like to have -rc2 tomorrow or on Friday.

That's good news!
Do you have a rough timeline for the final release? I would like to push a rc release into Arch's testing repository, but I do not want to block for too long...

**Comment by karelzak** [comment]
I'll be on holiday from March 7th to March 18th, and I'm uncertain about the speed of the translators (for .po files) in completing their work before March 7th. Therefore, the worst-case scenario is that the final release will be on March 19th.

**Comment by thesamesam** [comment]
We have systemd-tmpfiles available on OpenRC systems in Gentoo (and in the past, we had opentmpfiles too, which was a separate implementation). Could you add a way to install it there please (either via an option, or just install the tmpfiles unconditionally)?

**Comment by eworm-de** [comment]
Oh, did not know something like this exists... Perhaps we should revert the commit then, and also make the file for uuidd available.

**Comment by karelzak** [comment]
> Oh, did not know something like this exists... Perhaps we should revert the commit then, and also make the file for uuidd available.

I'd like to avoid revert in this case, it seems better to improve the code by a new PR.

What we need is to introduce systemd independent definition of `tmpfilesdir` in meson.build and ./configure (probably a hardcoded path if we cannot read it from  `pkg-config systemd`), and install the files also when systemd is not enabled.

I'd also supped to add --disable-tmpfiles for users who do not want it at all.

**Pull request merged**